### ISSUE TITLE
[Ask for help] Why are there two or more U-Nets in the model?

### ISSUE THREAD

**Issue by zzysay** [lucidrains/imagen-pytorch#152]
This may be a simple problem, but it puzzles the beginner of text2image.

Imagen has three key modules, which are respectively responsible for the generation of 64px image and the super-resolution of 256/1024px images. It seems that each module is composed of a U-Net.

So, does U-Net1 corresponds to the text2image module and U-Net2 corresponds to the super-resolution module of 64->256? **If I only need to learn a model to generate 64px images based on given texts, how many U-Nets do I need to train?**

Waiting for someone to answer my doubts ...

**Comment by Nodja** [comment]
The unets as described in the paper are completely independent and only connected at inference time. So to answer your question, you only have to train 1 unet if all you want is to generate 64x64 images from text.

Here's how the 3 unets relate.
| Unet # | Inputs | Output|
| - | - | - |
| Unet 1 | text embeds | 64x64 image |
| Unet 2 | text embeds + 64x64 image | 256x256 image |
| Unet 3 | text embeds + 256x256 image | 1024x1024 image |

The general intuition of what the unets are doing is this: Unet 1 generates the general structure of the image, unet 2 will use that structure and add the important details, unet 3 acts as a more general upscaler and just adjusts the fine details and cleans up the image. Note that none of this is strictly true and all unets may end up having a big impact on the final look of the image.

During training the upscaler unets are given the text and a downscaled version of the image in your dataset, not the output of the previous unet. Thus you can train each unet independently as all the inputs of the next unet are not dependent of the previous unet.

I don't believe google explained the reasoning of why they split the model into 3 discrete models. But I can guess. The first obvious reason is stability, having the problem split into smaller problems makes it easier to troubleshoot the network at design time. The second advantage is parallelism, since the unets are completely independent, you can easily fit it into a cluster of GPUs/TPUs that don't have huge amounts of RAM.

I should note that this repo does not perfectly reflect Google's paper perfectly. Notably it has many improvements from other papers added onto it, like the ability to do inpainting, conditioning on other images and [general improvements to training/sampling speed](https://github.com/lucidrains/imagen-pytorch#experimental) among a million other details.

I hope this clarifies things and I didn't confuse you further :P

**Comment by zzysay** [comment]
Thank you very very much. This completely solves my confusion!

**Comment by zhaobingbingbing** [comment]
@Nodja . Hi, you made a great explanation. I also have some questions about this repo. I remember in imagen, those 3 unets are trained separately, so we need to prepare 64*64,256*256,1024*1024  images separately for 3  unets separately? How can we obtain a huge number of 1024*1024 images. And  in this repo, it trained unet1 and unet2 together, the input are resized to 256*256, not 64*64, is that mean if we want to train 3 unets together, the input should be 1024*1024?

**Comment by Nodja** [comment]
> How can we obtain a huge number of 10241024 images.

Any way you like, easiest way is probably using [img2dataset](https://github.com/rom1504/img2dataset) on Laion 5B and filter by minimum size. It will wield very noisy results tho, so the subsets of Laion are probably better.

>so we need to prepare 6464,256256,10241024 images separately for 3 unets separately?

Depends, you can just get 1024x1024 images and pass them directly to imagen, it'll resize to the image_sizes hyperparam corresponding to the unet being trained. i.e. if you pass image_sizes= (64, 256) it'll resize images passed to unet1 to 64x64 and unet2 to 256x256. The resizing with this method will be fast but low quality (nearest neighbor iirc), alternatively you can resize it on the fly using a pytorch dataset, or even resize all the files beforehand. If you want high quality resizes I recommend you do it using a pytorch dataset, then setup preloading on your dataloader so your CPU resizes the images before they get sent to imagen.

> in this repo, it trained unet1 and unet2 together,

Don't know where you got this idea but that's false. The unets are completely independent until sampling.

**Comment by martinriven** [comment]
@zhaobingbingbing  yeah your understanding is right
